src/util/DeferredProperty.java

```diff
@@ -0,0 +1,231 @@
+     */
+    private final List<DeferredProperty<?>> observers;
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     * If no initial value is provided (i.e. <code>null</code> value) then this instance will start of as dirty.
+     * Otherwise, if a none <code>null</code> value is provided then it will start of clean.
+     *
+     * @param value The initial value of this property.
+     * @param supplier The non-parametric factory method for updating this property value automatically.
+     * @param observers The dependent observers to which marking this instance as dirty is propagated to.
+     */
+    public DeferredProperty(T value, Supplier<T> supplier, List<DeferredProperty<?>> observers) {
+        this.value = value;
+        this.supplier = supplier;
+        this.observers = observers;
+        isDirty = value == null;
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     * @param supplier The non-parametric factory method for updating this property value automatically.
+     */
+    public DeferredProperty(T value, Supplier<T> supplier) {
+        this(value, supplier, new ArrayList<>());
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     * @param observers The dependent observers to which marking this instance as dirty is propagated to.
+     */
+    public DeferredProperty(T value, List<DeferredProperty<?>> observers) {
+        this(value, null, observers);
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     * @param observers The dependent observers to which marking this instance as dirty is propagated to.
+     */
+    public DeferredProperty(T value, DeferredProperty<?> ... observers) {
+        this(value, List.of(observers));
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     */
+    public DeferredProperty(T value) {
+        this(value, new ArrayList<>());
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>} with an initial value of <code>null</code> and no observers.
+     */
+    public DeferredProperty() {
+        this(null, new ArrayList<>());
+    }
+
+    /**
+     * Bypasses the dirty checks and just returns the value of {@link DeferredProperty#value}.
+     *
+     * @return The current value of {@link DeferredProperty#value}.
+     */
+    public T getValue() {
+        return value;
+    }
+
+    /**
+     * Marks this instance as dirty and only if this instance was initially marked as dirty do we propagate this to its {@link DeferredProperty#observers}.
+     *
+     * @return <code>True</code> if we marked this instance as dirty.
```

Comment:
Since you didn't specify what `false` as a return value means, I would say, based on your provided text, that `false` means that the function did *not* mark this instance as dirty. Furthermore, my first reaction was that `marked` refers to the final state of the instance after this function, but looking at the return value it refers to the process of marking, especially changing the marking from clean to dirty.

Suggestion:
```suggestion
     * @return {@code True} if this instance changed its classification from clean to dirty, and {@code False} if nothing changed, i.e., this instance was already classified as dirty.
```